### PR TITLE
WASM: Add support for Subroutines

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
This `PR` adds support for `subroutines` in the `WebAssembly` Backend. 

Functions from `C` are currently not supported and are handled with the help of `is_unsupported_function()`.

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
This is a `WIP` (Work In Progress). The tests might not pass currently.

**Comment by certik** [comment]
Looks good so far.

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -133,9 +133,10 @@ RUN(NAME goto_01 LABELS gfortran llvm)
 RUN(NAME goto_02 LABELS gfortran llvm)
 RUN(NAME goto_03 LABELS gfortran)
 
-RUN(NAME subroutines_01 LABELS gfortran llvm cpp x86)
-RUN(NAME subroutines_02 LABELS gfortran llvm cpp x86)
-RUN(NAME subroutines_04 LABELS gfortran llvm)
+RUN(NAME subroutines_01 LABELS gfortran llvm cpp x86 wasm)
+RUN(NAME subroutines_02 LABELS gfortran llvm cpp x86 wasm)
+RUN(NAME subroutines_03 LABELS gfortran llvm wasm)
+RUN(NAME subroutines_04 LABELS gfortran llvm wasm)
```
Comment (Shaikh-Ubaid) [inline integration_tests/CMakeLists.txt:139]: `CPP/Kokkos` is not working on my system (It used to work previously, I am unsure why it stopped working), so I could not enable these tests for the `cpp` backend.

**Comment by Shaikh-Ubaid** [comment]
This is ready for review. Please possibly review and please share feedback.

**Comment by Shaikh-Ubaid** [comment]
I updated the description. 

This is ready for review. Please possibly review and please share feedback.

**Comment by Shaikh-Ubaid** [comment]
I just change a commit message slightly. This is ready for review.

**Review by certik — APPROVED** [review]
I think that this is fine.

What is the purpose of adding the print statements into the tests? I noticed it changes all the LLVM and ASR tests, so the PR diff becomes huge.

Do you need it for debugging?

It seems that the WAT test will test that the WASM was generated correctly, even without the print statements.

I think this is fine to merge as is.

**Comment by Shaikh-Ubaid** [comment]
```quote
What is the purpose of adding the print statements into the tests? 
I noticed it changes all the LLVM and ASR tests, so the PR diff becomes huge.

Do you need it for debugging?
```
They are added for the following benefits:
- We know that the generated `wasm` works if there is no `ERROR STOP` printed. At the same, explicitly printing intermediate output makes it more clear/stronger that the generated `wasm` works.
- I execute the same tests for `llvm` and `wasm` and try to observe if the outputs for both match exactly/closest-possible. The print statements help in observing if the outputs match.
- I currently use the `node 14.x` and it seems `i64` types are an experimental feature for `node 14.x` and therefore the command to execute should include `--experimental-wasm-bigint` flag. 
    - If the generated `wasm` (the `a.out` file) contains `i64` types and if the execution command is `node a.out.js` (that is, it is executed without the `--experimental-wasm-bigint` flag), there is nothing printed on the console (even if there are `print` statements used within the `source code`).
    - I guess, ideally, there should have been an error that says something like `Error: Experimental type i64, please pass --experimental-wasm-bigint flag`
    - Since, there is no error and at the same time there is no output (when using types `i64` and not using `--expeirmental-wasm-bigint`), I guess, this may/might be a bug with `i64` as it seems to be an experimental feature for the `node 14.x` that I use.
    - Just to be safe and to ensure that the generated `wasm` works fine, that is it prints the intermediate outputs, variables and types `i64` (with the appropriate flag `--experimental-wasm-bigint` passed), I added the `print` statements.

**Comment by Shaikh-Ubaid** [comment]
>I think this is fine to merge as is.

Got it. For the moment, I am merging this. If there are any suggestions, I will work on them in another `PR/PRs`.

Also, thank you so much for the approval.

**Comment by certik** [comment]
Got it. I figured you must have good reasons to put it in. I think that is fine, let's keep it there.

**Pull request merged**